SecretManagement.KeePass.Extension/Tests/Test-SecretVault.Tests.ps1

```diff
@@ -0,0 +1,60 @@
+get-module *Secret* | Remove-Module -ErrorAction SilentlyContinue
+Import-Module -name Microsoft.PowerShell.SecretManagement
+Microsoft.PowerShell.SecretManagement\Get-SecretVault -Name keepasspestertest* | Microsoft.PowerShell.SecretManagement\Unregister-SecretVault -erroraction SilentlyContinue
+Import-Module -Name "$($PSScriptRoot)\..\secretmanagement.keepass.extension.psd1" -force
+
+InModuleScope -ModuleName 'SecretManagement.KeePass.Extension' {
+    Describe "Test-SecretVault" {
+        BeforeAll {
+            $ModuleName = 'SecretManagement.Keepass.Extension'
+            $MasterKey = '"1}`.2R{LX1`Jm8%XX2/'
+            $KeepassDatabase = "Testdb.kdbx"
+    
+            $VaultName = "KeepassPesterTest_$([guid]::NewGuid())"
+            $VaultPath = Join-Path -Path $TestDrive -ChildPath $KeepassDatabase
+            Copy-Item -Path Testdb.kdbx -Destination $VaultPath
+            $VaultKey = [PSCredential]::new('vaultkey',(ConvertTo-SecureString -AsPlainText -Force $MasterKey))
+    
+            $RegisterSecretVaultParams = @{
+                Name            = $VaultName
+                ModuleName      = 'SecretManagement.Keepass'
+                PassThru        = $true
+                VaultParameters = @{
+                    Path = $VaultPath
+                }
+            }
+
+            Mock -Verifiable -CommandName 'Get-Credential' -MockWith {$VaultKey}
+        }
+        AfterAll {
+            try { $Vaults = Microsoft.PowerShell.SecretManagement\Get-SecretVault -name $VaultName } catch [System.Management.Automation.ItemNotFoundException] { }
+            if ($Vaults) { $Vaults | Microsoft.PowerShell.SecretManagement\Unregister-SecretVault}
+        }
+        Context "Validating Master Key rules" {
+            BeforeAll {
+                $TheVault = Microsoft.PowerShell.SecretManagement\Register-SecretVault @RegisterSecretVaultParams
+            }
+            AfterAll {
+                try { $Vaults = Microsoft.PowerShell.SecretManagement\Get-SecretVault -name $VaultName } catch [System.Management.Automation.ItemNotFoundException] { }
+                if ($Vaults) { $Vaults | Microsoft.PowerShell.SecretManagement\Unregister-SecretVault}
+            }
+            It "Should not have a variable 'Vault_$($VaultName)'" {
+                { (Get-Variable -Name "Vault_$VaultName" -Scope Script).Value } | Should -Throw
+            }
+            It "Should request a credential on the first pass" {
+                Test-SecretVault -VaultName $VaultName
+                Assert-MockCalled -CommandName 'Get-Credential' -Exactly 1
+            }
+            It 'Should not request a credential on the second pass' {
```

Comment:
I don't understand this test. You say "should not request" but your assertion is that it called Get-Credential meaning it did prompt.

Reply:
On the first pass of 'Test-Secret', the amount of times the cmdlet 'Get-Credential' is executed (mocked in the tests) is 1 because you want to request a credential.
From that point on, the 'Get-Credential' is no longer invoked anytime you run Test-SecretVault because your private variable Vault_.... alreay exists.
This tests describes exactly that: the number of times 'Get-Credential' is executed within the Context scope remains at 1

Reply:
OK, I think Assert-MockCalled is scoped within the It though isn't it? meaning that it'll say 1 because get-credential was called. I may be wrong there.

Reply:
Scope context makes this work as expected, thanks. However this test is not individually validatable, it requires the context of another test or it will fail if tested individually. I'll fix it later if it really annoys me :)

Reply:
I added it because it is a real life scenario:  you don't want to make an error in the Test-SecretVault function that will constantly request the credential information over and over again.

Reply:
Right, it makes sense as a test, however it should be structured all within a single "it" block. If I click "run test" on that test without running it on another it will fail because it requires previous state that isn't captured in a beforeeach/beforeall or mocked.